sound/soc/sof/sof-client-fw-gdb.c

```diff
@@ -0,0 +1,333 @@
+// SPDX-License-Identifier: GPL-2.0-only
+//
+// Copyright(c) 2022 Intel Corporation. All rights reserved.
+//
+// Author: Noah Klayman <noah.klayman@intel.com>
+//
+
+#include <linux/auxiliary_bus.h>
+#include <linux/completion.h>
+#include <linux/debugfs.h>
+#include <linux/ktime.h>
+#include <linux/mod_devicetable.h>
+#include <linux/module.h>
+#include <linux/pm_runtime.h>
+#include <linux/slab.h>
+#include <linux/uaccess.h>
+#include <sound/sof/header.h>
+#include <sound/sof/ipc4/header.h>
+
+#include "sof-client.h"
+#include "sof-priv.h"
+#include "ops.h"
+
+#define SOF_IPC_CLIENT_SUSPEND_DELAY_MS	3000
+
+struct sof_fw_gdb_priv {
+	struct dentry *dfs_file;
+	size_t max_msg_size;
+	enum sof_ipc_type ipc_type;
+
+	struct sof_ipc_gdb_message *tx_buffer;
+	void *rx_buffer;
+};
+
+struct sof_ipc_gdb_message {
+	struct sof_ipc_cmd_hdr hdr;
+	char cmd[];
+} __packed;
+
+#define RING_SIZE 128
+struct ring {
+	unsigned char head;
+	unsigned char fill1[63];
+	unsigned char tail;
+	unsigned char fill2[63];
+	unsigned char data[RING_SIZE];
+} __packed;
+
+static int sof_fw_gdb_send_message(struct sof_client_dev *cdev)
+{
+	struct sof_fw_gdb_priv *priv = cdev->data;
+	struct device *dev = &cdev->auxdev.dev;
+	int ret, err;
+
+	ret = pm_runtime_resume_and_get(dev);
```

Comment:
If we are debugging do we really want to support suspend/resume. If we are debugging a SOF module we should force device into D0.